worker/introspection/leases.go

```diff
@@ -5,36 +5,48 @@ package introspection
 
 import (
 	"fmt"
+	"math/rand"
 	"net/http"
 	"strings"
+	"sync/atomic"
 	"time"
 
+	"github.com/juju/errors"
+	"gopkg.in/yaml.v2"
+
 	"github.com/juju/juju/core/lease"
 	"github.com/juju/juju/core/raftlease"
-	"gopkg.in/yaml.v2"
 )
 
 type leaseHandler struct {
 	leases Leases
+	hub    StructuredHub
+	clock  Clock
+
+	runID     int32
+	requestID uint64
 }
 
-// ServeHTTP is part of the http.Handler interface.
-func (h leaseHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+func (h *leaseHandler) snapshot() (*raftlease.Snapshot, error) {
 	ss, err := h.leases.Snapshot()
 	if err != nil {
-		w.WriteHeader(http.StatusInternalServerError)
-		fmt.Fprintf(w, "error: %v\n", err)
-		return
+		return nil, errors.Annotate(err, "snapshot")
 	}
 	snapshot, ok := ss.(*raftlease.Snapshot)
 	if !ok {
-		w.WriteHeader(http.StatusInternalServerError)
-		fmt.Fprintf(w, "expected *raftlease.Snapshot\n")
-		return
+		return nil, errors.New("expected *raftlease.Snapshot")
 	}
 	if snapshot.Version != 1 {
+		return nil, errors.New("only understand how to show version 1 snapshots")
+	}
+	return snapshot, nil
+}
+
+func (h *leaseHandler) list(w http.ResponseWriter, r *http.Request) {
+	snapshot, err := h.snapshot()
+	if err != nil {
 		w.WriteHeader(http.StatusInternalServerError)
```

Comment:
Is it worth spending some time here to report better http status codes. My reasoning is errors like only support version 1 is 400 status code as the user hasn't used it correctly etc. Just a suggestion/thought.

Reply:
This is a StatusInternalServerError though. It's not bad client input.
We do return StatusBadRequest in other places where relevant.